File: agent-driven-architecture-main/examples/kotlin/block/escalation/src/main/kotlin/adr/blocks/escalation/Project.kt
// ── blocks/escalation/project — TWO pure projections of the SAME slice ─────
// SITE 2 OF 3 (the view row match) and SITE 3 OF 3 (the context line match) for a
// new TicketStatus variant (§11.2). Both are closed matches with NO else arm, so
// the compiler's edit list for `Archived(ticket, at)` is exactly these three sites,
// all inside blocks/escalation/. Zero sites outside the block.
//
// This is also G12's fix in place: `t.status is Open` as a way to compute
// `canEscalate` is NOT a closed match — it compiles happily after a variant is
// added and silently answers the wrong thing.

package adr.blocks.escalation

import adr.spine.pure.MAX_CONTEXT_LINES_PER_BLOCK

data class EscalationRow(
    val ticket: String,
    val state: String,
    /** Pre-decided here, not in the view. */
    val canEscalate: Boolean,
    val escalating: Boolean,
    val escalated: Boolean,
)

data class EscalationView(val rows: List<EscalationRow>)

internal class EscalationProjection {

    fun view(slice: EscalationSlice): EscalationView = EscalationView(
        rows = slice.status.values.map { status ->
            when (status) {
                is TicketStatus.Open -> EscalationRow(
                    ticket = status.ticket.value,
                    state = "open",
                    canEscalate = true,
                    escalating = false,
                    escalated = false,
                )

                is TicketStatus.Escalating -> EscalationRow(
                    ticket = status.ticket.value,
                    state = "escalating (asked by ${status.requestedBy.id})",
                    canEscalate = false,
                    escalating = true,
                    escalated = false,
                )

                is TicketStatus.Escalated -> EscalationRow(
                    ticket = status.ticket.value,
                    state = "escalated (confirmed by ${status.confirmedBy.id})",
                    canEscalate = false,
                    escalating = false,
                    escalated = true,
                )

                is TicketStatus.Resolved -> EscalationRow(
                    ticket = status.ticket.value,
                    state = "resolved",
                    canEscalate = false,
                    escalating = false,
                    escalated = false,
                )
            }
        },
    )

    fun contextLines(slice: EscalationSlice, maxLines: Int = MAX_CONTEXT_LINES_PER_BLOCK): List<String> =
        slice.status.values.take(maxLines).map { status ->
            when (status) {
                is TicketStatus.Open -> "ticket ${status.ticket.value} is open"
                is TicketStatus.Escalating ->
                    "ticket ${status.ticket.value} has a PENDING escalation request"

                is TicketStatus.Escalated -> "ticket ${status.ticket.value} is escalated; on-call paged"
                is TicketStatus.Resolved -> "ticket ${status.ticket.value} is resolved"
            }
        }
}

File: agent-driven-architecture-main/examples/typescript/src/blocks/escalation/project.ts
// ── blocks/escalation/project — CONSUMERS 2 and 3 of TicketStatus (G12) ────
// This is the file the shipped reference failed 15.4's own G12 self-check in:
//
//     canEscalate: t.status.kind === "Open",
//     escalating:  t.status.kind === "Escalating",
//
// Adding `Archived` to the status union left that compiling clean, exit 0, 8/8
// green — so the "the compiler hands you the edit list" promise was false at the
// one place a reader would check it. An `===` against a single variant is NOT a
// closed match.
//
// Below, both projections are `never`-guarded switches over every variant. Add
// a variant and the build breaks HERE, twice, plus once in `fold.ts` — three
// sites, all inside this block folder, zero outside it. Prove it yourself:
//   add `| Archived` to TicketStatus, run `npm run typecheck`, expect 3 errors.

import { bounded, MAX_CONTEXT_LINES_PER_BLOCK } from "@adr/spine/pure/context";
import type { EscalationSlice, TicketStatus } from "./slice";

export interface EscalationRow {
  readonly ticket: string;
  readonly status: string;
  readonly canEscalate: boolean;
  readonly escalating: boolean;
  readonly escalated: boolean;
}

export interface EscalationView {
  readonly rows: readonly EscalationRow[];
}

export function escalationView(slice: EscalationSlice): EscalationView {
  return { rows: [...slice.statuses.values()].map(rowFor) };
}

// CONSUMER 2 — the view row.
function rowFor(status: TicketStatus): EscalationRow {
  switch (status.kind) {
    case "Open":
      return {
        ticket: status.ticket,
        status: "open",
        canEscalate: true,
        escalating: false,
        escalated: false,
      };
    case "Escalating":
      return {
        ticket: status.ticket,
        status: "escalating",
        canEscalate: false,
        escalating: true,
        escalated: false,
      };
    case "Escalated":
      return {
        ticket: status.ticket,
        status: "escalated",
        canEscalate: false,
        escalating: false,
        escalated: true,
      };
    case "Resolved":
      return {
        ticket: status.ticket,
        status: "resolved",
        canEscalate: false,
        escalating: false,
        escalated: false,
      };
    default: {
      const _never: never = status;
      return _never;
    }
  }
}

export function escalationContextLines(
  slice: EscalationSlice,
  max: number = MAX_CONTEXT_LINES_PER_BLOCK,
): readonly string[] {
  const lines = [...slice.statuses.values()].map(contextLineFor);
  return bounded(lines, max);
}

// CONSUMER 3 — the reasoner's digest. A pure projection of the same slice, held
// to the same closed-match rule, so what the model reads and what the user sees
// can never drift apart silently.
function contextLineFor(status: TicketStatus): string {
  switch (status.kind) {
    case "Open":
      return `ticket ${status.ticket}: open, may be escalated`;
    case "Escalating":
      return `ticket ${status.ticket}: escalation requested, awaiting a different authority`;
    case "Escalated":
      return `ticket ${status.ticket}: escalated, on-call paged`;
    case "Resolved":
      return `ticket ${status.ticket}: resolved`;
    default: {
      const _never: never = status;
      return _never;
    }
  }
}
